### PR TITLE
AP_NavEKF3: add support for external navigation (aka vision-position-estimates)

### PR DESCRIPTION
**Pull request by rmackay9**
This PR adds support for EKF3 to accept external position estimates from a vicon or vision system like the T265.  This feature already exists in EKF2 so this is mostly a port of the EK2 feature to EKF3 with some changes to account for how the EKF3 accepts external yaw estimates.

I've tested this in SITL and it works.  A flight test on a real-vehicle will be done in the near future.

This is a replacement of PRs https://github.com/ArduPilot/ardupilot/pull/10374 and https://github.com/ArduPilot/ardupilot/pull/8730.

It might be best/fair to get some of @priseborough's PRs in first including the emergency yaw reset PR: https://github.com/ArduPilot/ardupilot/pull/13776

Resolved issues with this PR:

- [x] The external nav yaw does not seem to be being consumed.  When COMPASS_USE = 0 the EKF3 will not initialise it's yaw heading and if it's initialised by momentarily setting COMPASS_USE = 1 (and then switching it back to zero) the heading is not updated with the external nav's yaw's provided value. -- FIXED.  We need to set EK3_MAG_CAL = 5 (external yaw)
- [x] the Copter.VisionPositionEKF3 test is failing near the end of the test with a "Should not be called on invalid location" message displayed on the console.  The test mostly works, the vehicle flies using the external position estimate but when it comes time to land and disarm this failure message is displayed. -- FIXED.  The issue was related to the autotest changing between EKF2 and EKF3 which leads to a panic in SITL but this is not related to the change.
- [x] Position resets do not work correctly ([same issue as EKF2](https://github.com/ArduPilot/ardupilot/issues/14046)) -- FIXED
- [x] EKF3 is missing the EKx_EXTNAV_DELAY parameter but instead of adding this parameter to EKF3 I'm going to move the parameter to be part of the AP_VisualOdom library similar to how the AP_GPS library holds the expected GPS delay. -- This will be handled in a follow-up PR

### CODE REVIEW

**Comment by rmackay9** [comment]
I've resolved the position reset issue and re-tested in master and it seems to be working now.  Below is a picture of the Z-axis position reset working correctly.  The graph shows the VISP.PZ (up is negative) and then the POS.Alt (up is positive).  Horizontal resets also work.
image

**Comment by rmackay9** [comment]
rebased on master..

**Comment by rmackay9** [comment]
This is failing SITL because of issues fixed in PR https://github.com/ArduPilot/ardupilot/pull/14183 and PR https://github.com/ArduPilot/ardupilot/pull/14184

**Comment by rmackay9** [comment]
I've successfully flown this on a real vehicle now.  It seemed to work a-ok.

**Comment by rmackay9** [comment]
@priseborough, no pressure of course, I know you've been doing a lot of AP work recently but if you have time to review this sometime that would be greatly appreciated!

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 1]
```diff
@@ -369,13 +484,20 @@ void NavEKF3_core::SelectVelPosFusion()
...
+        if (activeHgtSource == HGT_SOURCE_EXTNAV) {
+            ResetPositionD(-hgtMea);
+        }
+    }
+
     // If we are operating without any aiding, fuse in the last known position
     // to constrain tilt drift. This assumes a non-manoeuvring vehicle
     // Do this to coincide with the height fusion
     if (fuseHgtData && PV_AidingMode == AID_NONE) {
-        gpsDataDelayed.vel.zero();
```
Comment (rmackay9) [inline libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:376]: @priseborough,

Do you think we should have "velPosObs[0,1,2] = 0" here to be the equivalent of gpsDataDelayed.vel.zero()?

Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:376]: It shouldn't be using GPS velocity in this mode so removing this line should make no difference.

File: libraries/AP_NavEKF3/AP_NavEKF3_core.h [hunk 2]
```diff
@@ -949,6 +982,7 @@ class NavEKF3_core : public NavEKF_core_common
     uint32_t airborneDetectTime_ms; // last time flight movement was detected
     Vector6 innovVelPos;            // innovation output for a group of measurements
     Vector6 varInnovVelPos;         // innovation variance output for a group of measurements
+    Vector6 velPosObs;              // observations for combined velocity and positon group of measurements (3x1 m , 3x1 m/s)
```
Comment (rmackay9) [inline libraries/AP_NavEKF3/AP_NavEKF3_core.h:985]: @tridge says we should declare this differently so we can test we are not running past the end of the array.

**Comment by priseborough** [comment]
https://github.com/ArduPilot/ardupilot/blob/05ef466b52ef0cccb83ae08b01388ca3e10ad5e5/libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp#L434 is going to cause problems because regardless of whther GPos has been inhibited via frontend->_fusionModeGPS == 3, it will reset states to GPs if two receivers are fitted and there is a receiver selection change.

File: libraries/AP_NavEKF3/AP_NavEKF3_core.h (hunk 2 above)
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.h:985]: I thought that was what that type supported if MATH_CHECK_INDEXES ?

Comment (rmackay9) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.h:985]: Yes, I think you're right that it's using the correct type, thanks.

**Comment by rmackay9** [comment]
I'll give this another hour for final comments and then I'll merge, thanks!

File: libraries/AP_NavEKF3/AP_NavEKF3_core.h (hunk 2 above)
Comment (tridge) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.h:985]: yes, Vector6 is good. I was referring to places in the EKF where we do things like this:
  float H_DECL[24] = {};
that bypasses the index checking

**Comment by priseborough** [comment]
OK with me

**Comment by rmackay9** [comment]
Merged!  thanks very much for the reviews!

**Pull request merged**